keras_nlp/models/gpt2/gpt2_causal_lm.py

```diff
@@ -361,6 +355,57 @@ def next(prompt, cache, index):
             hidden_states=hidden_states,
         )
 
+        # Compute an output padding mask with the token ids we updated.
+        if end_token_id is not None:
+            end_locations = (token_ids == end_token_id) & (~padding_mask)
```

Comment:
This is the first application of bitwise operations I've seen in python code. Super cool!

This block is a bit inscrutable at first glance however. Is there another comment or easier way to write this out? (Why do we need a cumsum?)

Reply:
Added some comments!

Reply:
Thank for the clarifications! I bet you're good at leetcode 😄